tests/src/main/java/org/hibernate/beanvalidation/tck/tests/constraints/containerelement/NestedContainerElementConstraintsTest.java

```diff
@@ -0,0 +1,297 @@
+import static org.hibernate.beanvalidation.tck.util.ConstraintViolationAssert.assertThat;
+import static org.hibernate.beanvalidation.tck.util.ConstraintViolationAssert.pathWith;
+import static org.hibernate.beanvalidation.tck.util.ConstraintViolationAssert.violationWith;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import javax.validation.ConstraintViolation;
+import javax.validation.constraints.NotNull;
+import javax.validation.constraints.Size;
+import javax.validation.valueextraction.Unwrapping;
+
+import org.hibernate.beanvalidation.tck.beanvalidation.Sections;
+import org.hibernate.beanvalidation.tck.tests.AbstractTCKTest;
+import org.jboss.arquillian.container.test.api.Deployment;
+import org.jboss.shrinkwrap.api.spec.WebArchive;
+import org.jboss.test.audit.annotations.SpecAssertion;
+import org.jboss.test.audit.annotations.SpecVersion;
+import org.testng.annotations.Test;
+
+import javafx.beans.property.SimpleStringProperty;
+import javafx.beans.property.StringProperty;
+
+@SuppressWarnings("restriction")
+@SpecVersion(spec = "beanvalidation", version = "2.0.0")
+public class NestedContainerElementConstraintsTest extends AbstractTCKTest {
+
+	@Deployment
+	public static WebArchive createTestArchive() {
+		return webArchiveBuilder()
+				.withTestClass( NestedContainerElementConstraintsTest.class )
+				.build();
+	}
+
+	@Test
+	@SpecAssertion(section = Sections.CONSTRAINTDECLARATIONVALIDATIONPROCESS_CONTAINERELEMENTCONSTRAINTS, id = "f")
+	public void validation_of_nested_type_arguments_works_with_map_of_list_of_optional() {
+		Set<ConstraintViolation<MapOfLists>> constraintViolations = getValidator().validate( MapOfLists.valid() );
+		assertNumberOfViolations( constraintViolations, 0 );
+
+		constraintViolations = getValidator().validate( MapOfLists.invalidKey() );
+		assertThat( constraintViolations ).containsOnlyViolations(
+				violationWith()
+						.constraintType( Size.class )
+						.propertyPath( pathWith()
+								.property( "map" )
+								.containerElement( "<map key>", true, "k", null, Map.class, 0 )
+						)
+		);
+
+		constraintViolations = getValidator().validate( MapOfLists.invalidList() );
+		assertThat( constraintViolations ).containsOnlyViolations(
+				violationWith()
+						.constraintType( Size.class )
+						.propertyPath( pathWith()
+								.property( "map" )
+								.containerElement( "<map value>", true, "key1", null, Map.class, 1 )
+						)
+		);
+
+		constraintViolations = getValidator().validate( MapOfLists.invalidString() );
+		assertThat( constraintViolations ).containsOnlyViolations(
+				violationWith()
+						.constraintType( Size.class )
+						.propertyPath( pathWith()
+								.property( "map" )
+								.containerElement( "<map value>", true, "key1", null, Map.class, 1 )
+								.containerElement( "<list element>", true, null, 0, List.class, 0 )
```

Comment:
Was wondering why there's no element for the `Optional`, but we don't add a node I'm realizing now.